crates/bevy_ecs/macros/src/fetch.rs

```diff
@@ -0,0 +1,622 @@
+            .map_or(false, |ident| ident == MUTABLE_ATTRIBUTE_NAME)
+    });
+
+    let world_lifetime = ast.generics.params.first().and_then(|param| match param {
+        lt @ GenericParam::Lifetime(_) => Some(lt.clone()),
+        _ => None,
+    });
+    // Fetch's HRTBs require substituting world lifetime with an additional one to make the
+    // implementation compile. I don't fully understand why this works though. If anyone's curious
+    // enough to try to find a better work around, I'll leave playground links here:
+    // - https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=da5e260a5c2f3e774142d60a199e854a (this fails)
+    // - https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=802517bb3d8f83c45ee8c0be360bb250 (this compiles)
+    let fetch_lifetime =
+        GenericParam::Lifetime(LifetimeDef::new(Lifetime::new("'fetch", Span::call_site())));
+
+    let has_world_lifetime = world_lifetime.is_some();
+    let world_lifetime = world_lifetime.unwrap_or_else(|| {
+        GenericParam::Lifetime(LifetimeDef::new(Lifetime::new("'world", Span::call_site())))
+    });
+    let state_lifetime =
+        GenericParam::Lifetime(LifetimeDef::new(Lifetime::new("'state", Span::call_site())));
+
+    let mut fetch_trait_punctuated_lifetimes = Punctuated::<_, Token![,]>::new();
+    fetch_trait_punctuated_lifetimes.push(world_lifetime.clone());
+    fetch_trait_punctuated_lifetimes.push(state_lifetime.clone());
+
+    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
+
+    let struct_name = ast.ident.clone();
+    let struct_name_read_only = if has_mutable_attr {
+        Ident::new(&format!("{}ReadOnly", struct_name), Span::call_site())
+    } else {
+        ast.ident.clone()
+    };
+    let fetch_struct_name = Ident::new(&format!("{}Fetch", struct_name), Span::call_site());
+    let state_struct_name = Ident::new(&format!("{}State", struct_name), Span::call_site());
+    let read_only_fetch_struct_name = if has_mutable_attr {
+        Ident::new(&format!("{}ReadOnlyFetch", struct_name), Span::call_site())
+    } else {
+        fetch_struct_name.clone()
+    };
+
+    let fields = match &ast.data {
+        Data::Struct(DataStruct {
+            fields: Fields::Named(fields),
+            ..
+        }) => &fields.named,
+        _ => panic!("Expected a struct with named fields"),
+    };
+
+    let mut phantom_field_idents = Vec::new();
+    let mut phantom_field_types = Vec::new();
+    let mut field_idents = Vec::new();
+    let mut field_types = Vec::new();
+    let mut query_types = Vec::new();
+    let mut fetch_init_types = Vec::new();
+
+    for field in fields.iter() {
+        let (world_lifetime, fetch_lifetime) = match (&world_lifetime, &fetch_lifetime) {
+            (GenericParam::Lifetime(world), GenericParam::Lifetime(fetch)) => {
+                (&world.lifetime, &fetch.lifetime)
+            }
+            _ => unreachable!(),
+        };
```

Comment:
This feels wrong inside the for loop, as it will be the same in every loop iteration.
But i also don't have any idea for another name, which would prevent shadowing when placed outside.

Reply:
`unpacked_world_lifetime` or `extracted_world_lifetime` perhaps? I agree that this should be out of the loop.

Reply:
Oh, I already went with `world_lifetime_param` for the original variable (as the wrapper type is called `GenericParam`) before seeing your suggestion. Wdyt?

Reply:
Excellent, I like that a lot better.